libkbfs/keybase_service_base.go

```diff
@@ -440,16 +448,58 @@ func (k *KeybaseServiceBase) LoadUserPlusKeys(ctx context.Context,
 // LoadTeamPlusKeys implements the KeybaseService interface for
 // KeybaseServiceBase.
 func (k *KeybaseServiceBase) LoadTeamPlusKeys(
-	ctx context.Context, tid keybase1.TeamID) (TeamInfo, error) {
+	ctx context.Context, tid keybase1.TeamID, desiredKeyGen KeyGen,
+	desiredUser keybase1.UserVersion, desiredRole keybase1.TeamRole) (
+	TeamInfo, error) {
+	if desiredRole != keybase1.TeamRole_NONE &&
+		desiredRole != keybase1.TeamRole_WRITER &&
+		desiredRole != keybase1.TeamRole_READER {
+		panic("KBFS shouldn't care about non-writer/reader roles")
+	}
+
 	cachedTeamInfo := k.getCachedTeamInfo(tid)
 	if cachedTeamInfo.Name != libkb.NormalizedUsername("") {
-		return cachedTeamInfo, nil
+		// If the cached team info doesn't satisfy our desires, don't
+		// use it.
+		satisfiesDesires := true
+		if desiredKeyGen >= FirstValidKeyGen {
+			satisfiesDesires = desiredKeyGen <= cachedTeamInfo.LatestKeyGen
+		}
+
+		if satisfiesDesires && desiredUser.Uid.Exists() {
+			// If the user is in the writer map, that satisfies none, reader
+			// or writer desires.
+			satisfiesDesires = cachedTeamInfo.Writers[desiredUser.Uid]
+			// If not, and the desire role is a reader, we need to
+			// check the reader map explicitly.
+			if !satisfiesDesires &&
+				(desiredRole == keybase1.TeamRole_NONE ||
+					desiredRole == keybase1.TeamRole_READER) {
+				satisfiesDesires = cachedTeamInfo.Readers[desiredUser.Uid]
```

Comment:
You could do less checking by doing the `cachedTeamInfo.Readers` check first, and then the writers check second.